libnetdata/gorilla/gorilla.cc

```diff
@@ -0,0 +1,618 @@
+
+    bit_code_init(&bcr, (Word *) src, src_len);
+
+    Word num_entries;
+    if (!bit_code_info(&bcr, &num_entries, (Word *) NULL)) {
+        return 0;
+    }
+    if (num_entries > dst_len) {
+        return false;
+    }
+    
+    for (size_t i = 0; i != num_entries; i++) {
+        if (!bit_code_read(&bcr, &dst[i]))
+            return 0;
+    }
+
+    return num_entries;
+}
+
+/*
+ * Low-level public API
+*/
+
+// 32-bit API
+
+void bit_code_writer_u32_init(bit_code_writer_u32_t *bcw, uint32_t *buffer, uint32_t capacity) {
+    bit_code_t<uint32_t> *bc = (bit_code_t<uint32_t> *) bcw;
+    bit_code_init(bc, buffer, capacity);
+}
+
+bool bit_code_writer_u32_write(bit_code_writer_u32_t *bcw, const uint32_t number) {
+    bit_code_t<uint32_t> *bc = (bit_code_t<uint32_t> *) bcw;
+    return bit_code_write(bc, number);
+}
+
+bool bit_code_writer_u32_flush(bit_code_writer_u32_t *bcw) {
+    bit_code_t<uint32_t> *bc = (bit_code_t<uint32_t> *) bcw;
+    return bit_code_flush(bc);
+}
+
+void bit_code_reader_u32_init(bit_code_reader_u32_t *bcr, uint32_t *buffer, uint32_t capacity) {
+    bit_code_t<uint32_t> *bc = (bit_code_t<uint32_t> *) bcr;
+    bit_code_init(bc, buffer, capacity);
+}
+
+bool bit_code_reader_u32_read(bit_code_reader_u32_t *bcr, uint32_t *number) {
+    bit_code_t<uint32_t> *bc = (bit_code_t<uint32_t> *) bcr;
+    return bit_code_read(bc, number);
+}
+
+bool bit_code_reader_u32_info(bit_code_reader_u32_t *bcr, uint32_t *num_entries_written,
+                                                          uint32_t *num_bits_written) {
+    bit_code_t<uint32_t> *bc = (bit_code_t<uint32_t> *) bcr;
+    return bit_code_info(bc, num_entries_written, num_bits_written);
+}
+
+// 64-bit API
+
+void bit_code_writer_u64_init(bit_code_writer_u64_t *bcw, uint64_t *buffer, uint64_t capacity) {
+    bit_code_t<uint64_t> *bc = (bit_code_t<uint64_t> *) bcw;
+    bit_code_init(bc, buffer, capacity);
+}
+
+bool bit_code_writer_u64_write(bit_code_writer_u64_t *bcw, const uint64_t number) {
+    bit_code_t<uint64_t> *bc = (bit_code_t<uint64_t> *) bcw;
+    return bit_code_write(bc, number);
+}
+
+bool bit_code_writer_u64_flush(bit_code_writer_u64_t *bcw) {
+    bit_code_t<uint64_t> *bc = (bit_code_t<uint64_t> *) bcw;
+    return bit_code_flush(bc);
+}
+
+void bit_code_reader_u64_init(bit_code_reader_u64_t *bcr, uint64_t *buffer, uint64_t capacity) {
+    bit_code_t<uint64_t> *bc = (bit_code_t<uint64_t> *) bcr;
+    bit_code_init(bc, buffer, capacity);
+}
+
+bool bit_code_reader_u64_read(bit_code_reader_u64_t *bcr, uint64_t *number) {
+    bit_code_t<uint64_t> *bc = (bit_code_t<uint64_t> *) bcr;
+    return bit_code_read(bc, number);
+}
+
+bool bit_code_reader_u64_info(bit_code_reader_u64_t *bcr, uint64_t *num_entries_written,
+                                                          uint64_t *num_bits_written) {
+    bit_code_t<uint64_t> *bc = (bit_code_t<uint64_t> *) bcr;
+    return bit_code_info(bc, num_entries_written, num_bits_written);
+}
```

Comment:
@vkalintiris both 32bit and 64bit functions have more or less same algorithm, but we are not using `function overload`, please, can you confirm we will expose this code for other languages? If it is not the case, it is not clear the motive we are not using `function overload`.

Reply:
I want to give @ktsaou the flexibility of picking whichever version he want from code that will be written in C.